pkg/sql/catalog/descs/hydrate.go

```diff
@@ -12,369 +12,165 @@ package descs
+			// was leased.
+			// TODO(ajwerner): Consider surfacing the mechanism used to retrieve the
+			// descriptor up to this layer.
+			if hd := desc.(catalog.HydratableDescriptor); tc.canUseHydratedDescriptorCache(hd.GetID()) {
+				if cached, err := tc.hydrated.GetHydratedDescriptor(ctx, hd, typeFn); err != nil {
+					return err
+				} else if cached != nil {
+					descs[i] = cached
+					continue
+				}
 			}
-
-			var err error
-			switch t := desc.(type) {
-			case catalog.TableDescriptor:
-				err = typedesc.HydrateTypesInTableDescriptor(ctx, t.TableDesc(), typedesc.TypeLookupFunc(typeLookup))
-			case catalog.FunctionDescriptor:
-				err = typedesc.HydrateTypesInFunctionDescriptor(ctx, t.FuncDesc(), typedesc.TypeLookupFunc(typeLookup))
-			case catalog.SchemaDescriptor:
-				err = typedesc.HydrateTypesInSchemaDescriptor(ctx, t.SchemaDesc(), typedesc.TypeLookupFunc(typeLookup))
-			}
-			if err != nil {
+			// The cache decided not to give back a hydrated descriptor, likely
+			// because either we've modified the table or one of the types or because
+			// this transaction has a stale view of one of the relevant descriptors.
+			// Proceed to hydrating a fresh copy.
+			desc = desc.NewBuilder().BuildImmutable()
+			if err := hydrateddesc.Hydrate(ctx, desc, typeFn); err != nil {
 				return err
 			}
+			descs[i] = desc
 		}
 	}
 	return nil
 }
 
-// hydrateCatalog installs type metadata in the types present for all
-// table descriptors in a catalog.
-func hydrateCatalog(ctx context.Context, cat nstree.Catalog) error {
-	// Since we have a catalog, we already have everything we need to hydrate our
-	// types. Set up an accessor for the type hydration method to look into the
-	// catalog.
-	typeLookup := func(ctx context.Context, id descpb.ID) (tree.TypeName, catalog.TypeDescriptor, error) {
-		desc := cat.LookupDescriptorEntry(id)
+func makeMutableTypeLookupFunc(
+	tc *Collection, txn *kv.Txn, descs []catalog.Descriptor,
+) typedesc.TypeLookupFunc {
+	var mut nstree.MutableCatalog
+	for _, desc := range descs {
 		if desc == nil {
-			n := tree.MakeUnresolvedName(fmt.Sprintf("[%d]", id))
-			return tree.TypeName{}, nil, sqlerrors.NewUndefinedObjectError(&n,
-				tree.TypeObject)
+			continue
 		}
-		typDesc, ok := desc.(catalog.TypeDescriptor)
-		if !ok {
-			return tree.TypeName{}, nil, errors.Newf(
-				"found %s while looking for type [%d]", desc.DescriptorType(), id,
-			)
+		if _, ok := desc.(catalog.MutableDescriptor); !ok {
+			continue
 		}
-		dbDesc := cat.LookupDescriptorEntry(typDesc.GetParentID())
-		if dbDesc == nil {
-			n := fmt.Sprintf("[%d]", typDesc.GetParentID())
-			return tree.TypeName{}, nil, sqlerrors.NewUndefinedDatabaseError(n)
+		mut.UpsertDescriptorEntry(desc)
+	}
+	mutableLookupFunc := func(ctx context.Context, id descpb.ID) (catalog.Descriptor, error) {
+		return tc.GetMutableDescriptorByID(ctx, txn, id)
+	}
+	return hydrateddesc.MakeTypeLookupFuncForHydration(mut.Catalog, mutableLookupFunc)
+}
```

Comment:
Here and below: more layering in which we query an in-memory data structure before doing a lookup. I haven't investigated into whether this was beneficial at all, it might not be. I don't think it matters either way.